lib/cockpit-components-listing.jsx

```diff
@@ -17,257 +17,267 @@
-                this.setState({ loadedTabs: loadedTabs, activeTab: tabIdx });
-            }
-            e.stopPropagation();
-            e.preventDefault();
-        },
-        render: function() {
-            var self = this;
-            // only enable navigation if a function is provided and the row isn't expanded (prevnt accidental navigation)
-            var allow_navigate = !!this.props.navigateToItem && !this.state.expanded;
-            var bodyProps = { className: '', onClick: this.handleClick };
-            var count_display = null;
+        e.stopPropagation();
+        e.preventDefault();
+    },
+    handleTabClick: function(tabIdx, e) {
+        // only consider primary mouse button
+        if (!e || e.button !== 0)
+            return;
+        var prevTab = this.state.activeTab;
+        var prevTabPresence = 'default';
+        var loadedTabs = this.state.loadedTabs;
+        if (prevTab !== tabIdx) {
+            // see if we need to unload the previous tab
+            if ('presence' in this.props.tabRenderers[prevTab])
+                prevTabPresence = this.props.tabRenderers[prevTab].presence;
 
-            var header_entries = this.props.columns.map(function(itm) {
-                if (typeof itm === 'string' || itm === null || itm instanceof String || React.isValidElement(itm))
-                    return (<td>{itm}</td>);
-                else if ('header' in itm && itm.header)
-                    return (<th>{itm.name}</th>);
-                else
-                    return (<td>{itm.name}</td>);
-            });
+            if (prevTabPresence == 'onlyActive')
+                delete loadedTabs[prevTab];
 
-            var expand_toggle = null;
-            if (this.props.tabRenderers.length > 0) {
-                expand_toggle = <td className="listing-ct-toggle" onClick={ allow_navigate?this.handleExpandClick:undefined }>
-                                    <i className="fa fa-fw"></i>
-                                </td>;
-            }
-            var listing_item_classes = ["listing-ct-item"];
-            if (!allow_navigate)
-                listing_item_classes.push("listing-ct-nonavigate");
-            var listing_item = (
-                <tr className={ listing_item_classes.join(' ') }
-                    onClick={ allow_navigate?this.handleNavigateClick:this.handleExpandClick }>
-                    {expand_toggle}
-                    {header_entries}
-                </tr>
-            );
+            // ensure the new tab is loaded and update state
+            loadedTabs[tabIdx] = true;
+            this.setState({ loadedTabs: loadedTabs, activeTab: tabIdx });
+        }
+        e.stopPropagation();
+        e.preventDefault();
+    },
+    render: function() {
+        var self = this;
+        // only enable navigation if a function is provided and the row isn't expanded (prevnt accidental navigation)
+        var allow_navigate = !!this.props.navigateToItem && !this.state.expanded;
```

Comment:
Since all the lines here are changed anyway, this would be a good chance to convert variables to camelCase.